app/domain/repository/persist_repository_from_upstream.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+class Repository::PersistRepositoryFromUpstream
+  def self.create_or_update_from_host_data(upstream_raw_host_data)
+    return unless upstream_raw_host_data
+
+    ActiveRecord::Base.transaction do
```

Comment:
Any particular reason to put all the finds & assignments in a db transaction?

Reply:
yeah, I'm curious if the transaction there was unnecessary -- there's only one INSERT or UPDATE happening here, and in the INSERT case it doesn't raise a RecordInvalid when it's not valid anyway, so it wouldn't rollback the transaction

Reply:
I initially kept this from the old code because the code read like it was uncertain of the existing data. With that in mind it made sense to make sure the find and update existed in the same transaction so that a name clash was found and updated/destroyed correctly in tandem with the targeted create/update here. That actually may not work that way with how transactions function. I'm wondering if removing the name clash logic out of here and doing an optimist update to the data instead would uncomplicate some of this.

Reply:
Discussed a bit offline and adding here - it looks like the clashing behavior was introduced around [here](https://github.com/librariesio/libraries.io/commit/5d2b4968590f63649392b488dff87bd5abd3b698#diff-efd45bdfac24905df71fad0af0d27a758f406f7dca1e151f2568bc0962bd9383L201-R207). It seems like the intention was to keep the repository in sync with what the upstream's knowledge of the repository is, based off the id. So if libraries has an existing repository with id: 111/name: foo, then the next request returns with id: 111/name: bar, then we'd delete the existing row for foo and replace it with new row bar. This would ensure that the entry mirrors what's in the upstream, vs an update that could potentially end up with a mix of attributes from the old and new repository data. 

I'm not really sure what the actual use case is in github/gitlab/bitbucket is/was though with same ids/changed names that also changes entire rows (maybe if the main branch gets force pushed with the contents of a totally different repo/fork of a different repo?), but that appears to be the intention AFAICT. 

Also to note from the brief archaeology splunking, it looks like the history of collecting the different repository datas:
1) started off as an async sidekiq job to poll the upstream every once in a while
2) moved into a centralized Repository.rb model
3) moved into individual concerns when there got to be too much if this_repo/elsif that_repo branching in all the methods
4) moved into where they are now as separate classes